pilot/pkg/controller/workloadentry/workloadentry_controller.go

```diff
@@ -93,32 +105,97 @@ func (c *Controller) Run(stop <-chan struct{}) {
 		go c.periodicWorkloadEntryCleanup(stop)
 		go c.cleanupQueue.Run(stop)
 	}
+
+	for i := 0; i < 5; i++ {
+		go wait.Until(c.worker, time.Second, stop)
+	}
+}
+
+func (c *Controller) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+type WorkType int
+
+const (
+	Register WorkType = iota
+	Unregister
+)
+
+type workItem struct {
+	entryName  string
+	workType   WorkType
+	proxy      *model.Proxy
+	conTime    time.Time
+	disConTime time.Time
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	item, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(item)
+
+	workItem, ok := item.(*workItem)
+	if !ok {
+		return true
+	}
+
+	switch workItem.workType {
+	case Register:
+		err := c.registerWorkload(workItem.entryName, workItem.proxy, workItem.conTime)
+		c.handleErr(err, item)
+	case Unregister:
+		err := c.unregisterWorkload(workItem.entryName, workItem.proxy, workItem.disConTime)
+		c.handleErr(err, item)
+	}
+	return true
 }
 
 func setConnectMeta(c *config.Config, controller string, conTime time.Time) {
 	c.Annotations[WorkloadControllerAnnotation] = controller
 	c.Annotations[ConnectedAtAnnotation] = conTime.Format(timeFormat)
+	delete(c.Annotations, DisconnectedAtAnnotation)
 }
 
-func (c *Controller) RegisterWorkload(proxy *model.Proxy, conTime time.Time) error {
+func (c *Controller) RegisterWorkload(proxy *model.Proxy, conTime time.Time) {
 	if !features.WorkloadEntryAutoRegistration || c == nil {
-		return nil
+		return
 	}
 	// check if the WE already exists, update the status
 	entryName := autoregisteredWorkloadEntryName(proxy)
 	if entryName == "" {
-		return nil
+		return
 	}
 
+	if err := c.registerWorkload(entryName, proxy, conTime); err != nil {
+		log.Errorf(err)
+		c.queue.AddRateLimited(&workItem{
+			entryName: entryName,
+			workType:  Register,
+			proxy:     proxy,
+			conTime:   conTime,
+		})
+	}
+	return
+}
+
+func (c *Controller) registerWorkload(entryName string, proxy *model.Proxy, conTime time.Time) error {
 	wle := c.store.Get(gvk.WorkloadEntry, entryName, proxy.Metadata.Namespace)
 	if wle != nil {
+		lastConTime, _ := time.Parse(timeFormat, wle.Annotations[ConnectedAtAnnotation])
+		// the proxy has reconnected to another pilot, not belong to this one.
+		if conTime.Before(lastConTime) {
+			return nil
+		}
 		// Try to patch, if it fails then try to create
 		_, err := c.store.Patch(*wle, func(cfg config.Config) config.Config {
```

Comment:
The `Patch` here will have the same problem described in #29534. If QueueUnregister happens concurrently, it will unset `ControllerAnnotation`. The patch will attempt to use `"op": "replace"` but will now require `"op": "add"`.

Reply:
`QueueUnregister ` now donot unset `WorkloadControllerAnnotation`, because when retry it needs to check this.